test/metabase/lib/swap_test.cljc

```diff
@@ -0,0 +1,123 @@
+
+(defspec swap-clauses-on-breakouts-test-permutations
+  (swapped-clauses-prop
+    lib/breakouts :breakout
+    (as-> (lib/query meta/metadata-provider (meta/table-metadata :orders)) $q
+      (lib/breakout $q (meta/field-metadata :products :category))
+      (lib/breakout $q (meta/field-metadata :people :source))
+      (lib/breakout $q (lib/with-binning
+                         (meta/field-metadata :orders :subtotal)
+                         (second (lib/available-binning-strategies $q (meta/field-metadata :orders :subtotal)))))
+      (lib/breakout $q (lib/with-temporal-bucket (meta/field-metadata :orders :created-at) :month)))))
+
+(defspec swap-clauses-on-filters-test-permutations
+  (swapped-clauses-prop
+    lib/filters :filters
+    (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
+        (lib/filter (lib/= (meta/field-metadata :products :category) "Doohickey"))
+        (lib/filter (lib/< (meta/field-metadata :products :created-at) "2024-01-01T00:00:00"))
+        (lib/filter (lib/is-null (meta/field-metadata :orders :discount)))
+        (lib/filter (lib/!= (meta/field-metadata :people :source) "Facebook")))))
+
+(defspec swap-clauses-on-expressions-test-permutations
+  (swapped-clauses-prop
+    lib/expressions :expressions
+    (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
+        (lib/expression "clean-discount" (lib/coalesce (meta/field-metadata :orders :discount) 0))
+        (lib/expression "discount fraction" (lib// (meta/field-metadata :orders :discount)
+                                                   (meta/field-metadata :orders :subtotal)))
+        (lib/expression "order month" (lib/get-month (meta/field-metadata :orders :created-at)))
+        (lib/expression "signup year" (lib/get-year  (meta/field-metadata :people :created-at))))))
+
+(defspec swap-clauses-on-joins-test-permutations
+  (swapped-clauses-prop
+    lib/joins :joins
+    (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
+        (lib/join (lib/join-clause (meta/table-metadata :people)
+                                   [(lib/= (meta/field-metadata :orders :user-id)
+                                           (meta/field-metadata :people :id))]))
+        (lib/join (lib/join-clause (meta/table-metadata :products)
+                                   [(lib/= (meta/field-metadata :orders :product-id)
+                                           (meta/field-metadata :products :id))]))
+        (lib/join (lib/join-clause (meta/table-metadata :checkins)
+                                   [(lib/= (meta/field-metadata :checkins :user-id)
+                                           (meta/field-metadata :orders :user-id))]))
+        (lib/join (lib/join-clause (meta/table-metadata :venues)
+                                   [(lib/= (meta/field-metadata :people :latitude)
+                                           (meta/field-metadata :venues :latitude))])))))
```

Comment:
Would swapping this with the join to `:people` break the query? I don't see anything that would prevent that.

Reply:
Joins have been dropped.